File: content/ressources/screenshot.py
import yaml
import glob
import os

import asyncio
from pyppeteer import launch, errors

from slugify import slugify

async def main():
    for filename in glob.iglob("**/*.yml", recursive=True):
        with open(filename, "r") as stream:
            content = yaml.safe_load(stream)
            category = content["category"]
            slug = slugify(content["name"])
            print(content.get("url", False))
            directory = f"{category}/screenshots/"
            if not os.path.exists(directory):
                os.makedirs(directory)
            imagename = f"{directory}/{slug}.png"
            if content.get("url", False) and not os.path.isfile(imagename):
                browser = await launch({"headless": True, "ignoreHTTPSErrors": True})
                page = await browser.newPage()
                try:
                    await page.goto(
                        content["url"],
                        {"waitUntil": "networkidle2", "ignoreHTTPSErrors": True},
                        timeout=10000
                    )
                except (errors.TimeoutError, errors.PageError):
                    print('Timeout')
                else:
                    await page.screenshot({"path": imagename})
                    await browser.close()
            elif os.path.isfile(imagename):
                print(imagename, ' exists.')


asyncio.get_event_loop().run_until_complete(main())

File: src/functions/screenshot.js
const puppeteer = require('puppeteer');
const fs = require('fs-extra');
const path = require('path');

const makeScreenshot = async (urls, paths) => {
  const browser = await puppeteer.launch({
    ignoreHTTPSErrors: true,
    headless: true,
    args: ['--enable-features=NetworkService'],
  });
  await Promise.all(
    urls.map(async (url, index) => {
      const savePath = paths[index];
      if (!url) {
        console.log('Url null. ', url, savePath);
        return true;
      }

      if (fs.existsSync(savePath)) {
        console.log('File exists. Aborting.', savePath);
        return true;
      }
      fs.mkdir(path.dirname(savePath), { recursive: true }, (err) => {
        console.log(err);
      });

      try {
        const page = await browser.newPage();
        await page.goto(url);
        await page.screenshot({ path: savePath });
      } catch (error) {
        console.log('Error.', error, url);
      }
      return true;
    }),
  );
  await browser.close();
  return true;
};

module.exports = makeScreenshot;
